src/lib/matchers.rs

```diff
@@ -163,6 +163,14 @@ impl<T: Clone + Debug + Eq + 'static> MappingStore<T> {
         self.get_to(from).map_or(false, |x| x == *to)
     }
 
+    /// Compute whether two sub-trees are isomorphic, based on their hashes.
+    ///
+    /// Isomorphism in this case is defined has having the same hash and the
+    /// same hash serialization.
+    pub fn is_isomorphic_hash(&self, from: NodeId<FromNodeId>, to: NodeId<ToNodeId>) -> bool {
```

Comment:
I assume we only use this function in places where we want check whether two things *aren't* equal? Even then, two things which hash to the same value might be distinct. So I'm a bit nervous about this function.

Reply:
It seems odd, I agree, but this is as specified in the GT code:

```java
    @Override
    public boolean isIsomorphicTo(ITree tree) {
        if (this.getHash() != tree.getHash())
            return false;
        else
            return this.toStaticHashString().equals(tree.toStaticHashString());
    }
```

Reply:
I'm about 99.9% certain that code is incorrect and works only because hash collisions are somewhat rare. But relying on hashes for correctness is a bad idea: they're an optimisation, not a guarantee.

Reply:
I don't disagree (and we already have an exact isomorphism test that we use in integration tests), but this is the way GT does it, so I tend to think we should start with this and see what happens in practice, rather than deviating from their implementation before we can do any testing. I'll raise an issue.

Reply:
Can we at least add a `debug_assert` which fails if a) the hashes are the same but b) the contents aren't? Otherwise this will get lost and forgotten until one day it bites us on the rear end.